File: packages/cli/src/commands/convert.ts
import { writeFile } from "node:fs/promises";
import path from "node:path";
import {
  CoMotionError,
  CoMotionNotFoundError,
  generateElementId,
  listPresentationEntries,
  normaliseSlideSvg,
  readPresentationFile,
  resolveWorkDir,
  type ProjectJson,
} from "@co-motion/core";
import type { CommandHandler } from "../registry.js";

/**
 * `co-motion convert <presentation-id>` — the author-initiated conversion
 * of a presentation's slides into the compliant container form (ADR-0012).
 *
 * ## Why the file-writing half lives here and not in core
 *
 * The maths and the normalisation itself are in `@co-motion/core`
 * (`slide/normalise.ts`), where the front end can reach them. That module
 * is required to stay free of Node built-ins (跨波接縫 2 for this wave),
 * and reading and writing files is exactly what a Node built-in is for. So
 * the split is: core decides what the bytes should become, and this command
 * is the one place in the codebase that puts them on disk.
 *
 * ## Never half-converted
 *
 * A presentation where three slides are in the container form and the
 * fourth is not is a worse state than the one the author started in, so
 * this command guards both ways it could happen:
 *
 * - Every slide is read and normalised BEFORE anything is written, so one
 *   unconvertible slide aborts the run with not a single byte written.
 * - The write loop keeps each slide's original bytes and rolls the already-
 *   written slides back if a later write fails (see `rollBack`), so a disk
 *   that accepts three writes and refuses the fourth does not leave a
 *   mixed-format presentation behind either.
 *
 * ## Conversion is not something that happens to you
 *
 * AC 2: opening a presentation must never rewrite it. Structurally, not by
 * discipline — `openPresentation`, `readPresentationFile` and
 * `setElementText` have no path to this module at all; this command is the
 * only caller of `normaliseSlideSvg` that writes anything.
 */

export interface ConvertInput {
  id: string;
}

export interface ConvertSlideOutcome {
  slidePath: string;
  changed: boolean;
  /** How many containers this slide gained. */
  wrapped: number;
}

export interface ConvertReport {
  slides: ConvertSlideOutcome[];
}

export type ConvertData = ConvertReport;

/** One slide, read and normalised, waiting for the write phase. */
interface PendingSlide {
  slidePath: string;
  /** Real filesystem path. Internal only — never surfaced in any message (ADR-0004). */
  realPath: string;
  /** The normalised document. */
  svg: string;
  /** The bytes on disk before this run, kept so a failed write can be rolled back. */
  original: string;
  wrapped: number;
}

export const convertCommand: CommandHandler<ConvertInput, ConvertData> = async (input) => {
  const report = await convertPresentationSlides(input.id);
  const changed = report.slides.filter((slide) => slide.changed).length;
  const untouched = report.slides.length - changed;
  return { ok: true, data: report, message: describe(changed, untouched) };
};

function describe(changed: number, untouched: number): string {
  if (changed === 0 && untouched === 0) return "沒有投影片需要轉換";
  if (untouched === 0) return `已轉換 ${changed} 張投影片`;
  return `已轉換 ${changed} 張投影片，${untouched} 張原本就合規`;
}

/**
 * Normalises every slide listed in the presentation's `project.json` and
 * writes the results back. The only function in the codebase that persists
 * a normalisation.
 */
export async function convertPresentationSlides(id: string): Promise<ConvertReport> {
  const workDir = await resolveWorkDir(id);
  const project = await readProject(id);

  const pending: PendingSlide[] = [];
  for (const slidePath of project.slides) {
    const original = await readPresentationFile(id, slidePath);
    const realPath = await resolveRealPath(id, workDir, slidePath);
    let result;
    try {
      result = normaliseSlideSvg(original, { generateId: generateElementId });
    } catch (error) {
      if (error instanceof CoMotionError) {
        throw new CoMotionError(`${slidePath}：${error.message}整份簡報都沒有被修改。`);
      }
      throw error;
    }
    pending.push({ slidePath, realPath, svg: result.svg, original, wrapped: result.wrapped });
  }

  const slides: ConvertSlideOutcome[] = [];
  /** Slides already written this run, newest last — the undo log for a mid-write failure. */
  const written: typeof pending = [];

  for (const slide of pending) {
    const changed = slide.svg !== slide.original;
    if (changed) {
      try {
        await writeFile(slide.realPath, slide.svg, "utf-8");
      } catch {
        // slide.realPath is a real filesystem path (ADR-0004) — never quote it.
        throw new CoMotionError(await rollBack(written, slide.slidePath));
      }
      written.push(slide);
    }
    slides.push({ slidePath: slide.slidePath, changed, wrapped: slide.wrapped });
  }
  return { slides };
}

/**
 * Puts back every slide this run had already rewritten before one of them
 * failed to write, and returns the message describing what actually
 * happened on disk.
 *
 * Reading and normalising every slide up front (the first loop) makes an
 * unconvertible slide harmless — nothing is written at all. It does NOT
 * make a failing *write* harmless: the disk is a shared, mutable thing that
 * can refuse the fourth write after accepting three, which would leave the
 * presentation half in the container form and half not. Gate round 1
 * reproduced exactly that with a read-only `slides/004.svg`.
 *
 * Restoring is best-effort, because the same disk that just refused a write
 * can refuse the restore too. So the message is assembled from what was
 * actually observed, never from what was intended: only a fully successful
 * rollback is allowed to claim the presentation is untouched, and a partial
 * one names the slides left in the new format. A reassuring message that
 * does not match the disk is worse than no message — it sends the author
 * looking for the problem in the wrong place.
 */
async function rollBack(written: readonly PendingSlide[], failedSlidePath: string): Promise<string> {
  const notRestored: string[] = [];
  for (const slide of written) {
    try {
      await writeFile(slide.realPath, slide.original, "utf-8");
    } catch {
      notRestored.push(slide.slidePath);
    }
  }
  const failure = `寫入投影片時發生錯誤：${failedSlidePath}。`;
  if (notRestored.length === 0) {
    return `${failure}整份簡報都沒有被修改。`;
  }
  return (
    `${failure}已改寫的投影片還原失敗，這幾張現在是轉換後的格式，其餘維持原樣：` +
    `${notRestored.join("、")}。請修復磁碟問題後重新執行 convert。`
  );
}

async function readProject(id: string): Promise<ProjectJson> {
  const raw = await readPresentationFile(id, "project.json");
  let project: ProjectJson;
  try {
    project = JSON.parse(raw) as ProjectJson;
  } catch {
    throw new CoMotionError("簡報設定檔已損毀");
  }
  if (!Array.isArray(project.slides)) {
    throw new CoMotionError("簡報設定檔已損毀：沒有 slides 清單");
  }
  return project;
}

/**
 * Turns a virtual slide path into the real file to write, without ever
 * joining caller-supplied text onto the work directory: every segment is
 * first proven to exist by listing its parent, so the path is assembled out
 * of names the filesystem itself produced (ADR-0004, third layer — the same
 * structural rule `virtual-fs.ts` enforces on the read side).
 */
async function resolveRealPath(id: string, workDir: string, virtualPath: string): Promise<string> {
  const segments = virtualPath.split("/").filter((segment) => segment.length > 0);
  if (segments.length === 0) {
    throw new CoMotionNotFoundError(`找不到檔案：${virtualPath}`);
  }
  let realPath = workDir;
  let walked = "";
  for (const segment of segments) {
    const entries = await listPresentationEntries(id, walked);
    const found = entries.find((entry) => entry === segment);
    if (found === undefined) {
      throw new CoMotionNotFoundError(`找不到檔案：${virtualPath}`);
    }
    realPath = path.join(realPath, found);
    walked = walked ? `${walked}/${found}` : found;
  }
  return realPath;
}

File: crates/comotion/src/commands/convert.rs
//! `comotion convert <presentation-id>`, ported from
//! `packages/cli/src/commands/convert.ts`. Every slide is read and
//! normalised BEFORE anything is written, so one unconvertible slide aborts
//! the run with not a single byte written; the write loop rolls already-
//! written slides back if a later write fails. Never goes through
//! `workspace::write` (does not occupy an undo step — TS writes directly,
//! and this ports that exactly).

use crate::errors::{CoMotionError, CoMotionResult};
use crate::result::{CommandResult, FailureKind};
use crate::slide::normalise::normalise_slide_svg;
use crate::workspace::{self, project::read_project_json, virtual_fs};
use crate::{argv, id};
use std::path::PathBuf;

pub fn run(args: &[String]) -> CommandResult {
    let presentation_id = match argv::require_id_positional(args, 0, "convert", "presentation-id") {
        Ok(v) => v,
        Err(msg) => return CommandResult::failure(msg, FailureKind::Failed),
    };
    match convert_presentation_slides(&presentation_id) {
        Ok(outcomes) => {
            let changed = outcomes.iter().filter(|o| o.changed).count();
            let untouched = outcomes.len() - changed;
            let data = serde_json::json!({
                "slides": outcomes.iter().map(|o| serde_json::json!({
                    "slidePath": o.slide_path,
                    "changed": o.changed,
                    "wrapped": o.wrapped,
                })).collect::<Vec<_>>(),
            });
            CommandResult::success(describe(changed, untouched), Some(data))
        }
        Err(err) => CommandResult::failure(err.message().to_string(), failure_kind_for(&err)),
    }
}

fn describe(changed: usize, untouched: usize) -> String {
    if changed == 0 && untouched == 0 {
        return "沒有投影片需要轉換".to_string();
    }
    if untouched == 0 {
        return format!("已轉換 {changed} 張投影片");
    }
    format!("已轉換 {changed} 張投影片，{untouched} 張原本就合規")
}

struct SlideOutcome {
    slide_path: String,
    changed: bool,
    wrapped: usize,
}

struct PendingSlide {
    slide_path: String,
    real_path: PathBuf,
    svg: String,
    original: String,
    wrapped: usize,
}

fn convert_presentation_slides(id_str: &str) -> CoMotionResult<Vec<SlideOutcome>> {
    let work_dir = workspace::resolve_work_dir(id_str)?;
    let project = read_project_json(&work_dir)?;

    let mut pending: Vec<PendingSlide> = Vec::with_capacity(project.slides.len());
    for slide_path in &project.slides {
        let original = virtual_fs::read_virtual_file(&work_dir, slide_path)?;
        let real_path = virtual_fs::resolve_virtual_file_path(&work_dir, slide_path)?;
        let mut generate_id = || format!("el-{}", id::generate_opaque_id());
        let result = normalise_slide_svg(&original, &mut generate_id).map_err(|err| {
            CoMotionError::invalid(format!(
                "{slide_path}：{}整份簡報都沒有被修改。",
                err.message()
            ))
        })?;
        pending.push(PendingSlide {
            slide_path: slide_path.clone(),
            real_path,
            svg: result.svg,
            original,
            wrapped: result.wrapped,
        });
    }

    let mut outcomes = Vec::with_capacity(pending.len());
    let mut written: Vec<&PendingSlide> = Vec::new();
    for slide in &pending {
        let changed = slide.svg != slide.original;
        if changed {
            if std::fs::write(&slide.real_path, &slide.svg).is_err() {
                return Err(CoMotionError::invalid(roll_back(&written, slide)));
            }
            written.push(slide);
        }
        outcomes.push(SlideOutcome {
            slide_path: slide.slide_path.clone(),
            changed,
            wrapped: slide.wrapped,
        });
    }
    Ok(outcomes)
}

/// Puts back every slide this run had already rewritten before one of them
/// failed to write, and returns the message describing what actually
/// happened on disk.
fn roll_back(written: &[&PendingSlide], failed: &PendingSlide) -> String {
    let mut not_restored: Vec<String> = Vec::new();
    for slide in written {
        if std::fs::write(&slide.real_path, &slide.original).is_err() {
            not_restored.push(slide.slide_path.clone());
        }
    }
    if !restored_to_original(failed) {
        not_restored.push(failed.slide_path.clone());
    }
    let failure = format!("寫入投影片時發生錯誤：{}。", failed.slide_path);
    if not_restored.is_empty() {
        return format!("{failure}整份簡報都沒有被修改。");
    }
    format!(
        "{failure}已改寫的投影片還原失敗，這幾張現在不是原始內容，其餘維持原樣：{}。請修復磁碟問題後重新執行 convert。",
        not_restored.join("、")
    )
}

fn restored_to_original(slide: &PendingSlide) -> bool {
    let current = match std::fs::read_to_string(&slide.real_path) {
        Ok(c) => c,
        Err(_) => return false,
    };
    if current == slide.original {
        return true;
    }
    std::fs::write(&slide.real_path, &slide.original).is_ok()
}

fn failure_kind_for(err: &CoMotionError) -> FailureKind {
    match err {
        CoMotionError::NotFound(_) => FailureKind::NotFound,
        CoMotionError::InvalidRequest(_) => FailureKind::Failed,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::workspace::registry;
    use std::path::PathBuf;

    fn temp_dir(label: &str) -> PathBuf {
        let dir = std::env::temp_dir().join(format!(
            "comotion-test-convert-{label}-{}",
            id::random_hex_suffix()
        ));
        std::fs::create_dir_all(&dir).unwrap();
        dir
    }

    struct Fixture {
        home: PathBuf,
        work: PathBuf,
        id: String,
        _guard: std::sync::MutexGuard<'static, ()>,
    }

    impl Fixture {
        fn new(label: &str, project_json: &str, slides: &[(&str, &str)]) -> Self {
            let guard = registry::ENV_LOCK.lock().unwrap();
            let home = temp_dir(&format!("{label}-home"));
            let work = temp_dir(&format!("{label}-work"));
            let test_id = format!("test-{label}");
            unsafe {
                std::env::set_var("COMOTION_HOME", &home);
            }
            std::fs::create_dir_all(work.join("slides")).unwrap();
            std::fs::write(work.join("project.json"), project_json).unwrap();
            for (path, content) in slides {
                std::fs::write(work.join(path), content).unwrap();
            }
            registry::register_for_test(&home, &test_id, &work);
            Fixture {
                home,
                work,
                id: test_id,
                _guard: guard,
            }
        }
    }

    impl Drop for Fixture {
        fn drop(&mut self) {
            unsafe {
                std::env::remove_var("COMOTION_HOME");
            }
            std::fs::remove_dir_all(&self.home).ok();
            std::fs::remove_dir_all(&self.work).ok();
        }
    }

    #[test]
    fn missing_id_argument_fails() {
        let result = run(&[]);
        assert!(!result.ok);
        assert_eq!(result.message, "命令 convert 缺少參數：presentation-id");
    }

    #[test]
    fn already_compliant_slide_reports_no_slides_need_conversion_when_empty() {
        let fixture = Fixture::new(
            "empty",
            r#"{"formatVersion":1,"name":"T","canvas":{"width":100,"height":100},"slides":[],"fonts":[]}"#,
            &[],
        );
        let result = run(&[fixture.id.clone()]);
        assert!(result.ok);
        assert_eq!(result.message, "沒有投影片需要轉換");
    }

    #[test]
    fn wraps_a_bare_primitive_and_reports_changed() {
        let fixture = Fixture::new(
            "wrap",
            r#"{"formatVersion":1,"name":"T","canvas":{"width":100,"height":100},"slides":["slides/001.svg"],"fonts":[]}"#,
            &[(
                "slides/001.svg",
                "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\">\n  <rect width=\"1\" height=\"1\"/>\n</svg>",
            )],
        );
        let result = run(&[fixture.id.clone()]);
        assert!(result.ok);
        assert_eq!(result.message, "已轉換 1 張投影片");
        let content = std::fs::read_to_string(fixture.work.join("slides/001.svg")).unwrap();
        assert!(content.contains("<g id="));
    }

    #[test]
    fn already_compliant_slide_is_untouched_and_reported_as_such() {
        let fixture = Fixture::new(
            "compliant",
            r#"{"formatVersion":1,"name":"T","canvas":{"width":100,"height":100},"slides":["slides/001.svg"],"fonts":[]}"#,
            &[(
                "slides/001.svg",
                "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\"><g id=\"a\"><rect width=\"1\" height=\"1\"/></g></svg>",
            )],
        );
        let before = std::fs::read_to_string(fixture.work.join("slides/001.svg")).unwrap();
        let result = run(&[fixture.id.clone()]);
        assert!(result.ok);
        assert_eq!(result.message, "已轉換 0 張投影片，1 張原本就合規");
        let after = std::fs::read_to_string(fixture.work.join("slides/001.svg")).unwrap();
        assert_eq!(before, after);
    }

    #[test]
    fn blocking_issue_writes_nothing_and_names_the_slide() {
        let fixture = Fixture::new(
            "blocking",
            r#"{"formatVersion":1,"name":"T","canvas":{"width":100,"height":100},"slides":["slides/001.svg"],"fonts":[]}"#,
            &[(
                "slides/001.svg",
                "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\"><script>evil()</script></svg>",
            )],
        );
        let before = std::fs::read_to_string(fixture.work.join("slides/001.svg")).unwrap();
        let result = run(&[fixture.id.clone()]);
        assert!(!result.ok);
        assert!(result.message.starts_with("slides/001.svg："));
        assert!(result.message.ends_with("整份簡報都沒有被修改。"));
        let after = std::fs::read_to_string(fixture.work.join("slides/001.svg")).unwrap();
        assert_eq!(before, after);
    }
}
